casper/src/main/resources/RevVault.rho

```diff
@@ -0,0 +1,154 @@
+// 4.  | 1,         | ed25519            | pk = f59e613088fd7e2e8a371b5115701f3a4f53f35ea6f60433923153ee7fbf83fd
+// 5.  | 4, 2,      | genIds             | uname = Unforgeable(0x5df3b04a7cc9aaa9dad1afa7b9bc9b1df9d6ccc76c5dff3c4d2d89b1ba6a3e8b)
+// 6.  | 3, 5,      | registry           | value = (9223372036854775807, bundle+ {   Unforgeable(0x5df3b04a7cc9aaa9dad1afa7b9bc9b1df9d6ccc76c5dff3c4d2d89b1ba6a3e8b) })
+// 7.  | 6,         | protobuf           | toSign = 2a3eaa013b0a0d2a0b10feffffffffffffffff010a2a5a280a243a220a205df3b04a7cc9aaa9dad1afa7b9bc9b1df9d6ccc76c5dff3c4d2d89b1ba6a3e8b1001
+// 8.  | 7, 1,      | ed25519            | sig = cd16cb789f128767d2d1224b931d0d5d47dce130266c72839d09d38bc9070a365234e0d9cce4a04e7ac4a1648db1b2ec1ba55751d0ead2c465dbfb55a5d13f01
+// 9.  | 4,         | registry           | uri = rho:id:1o93uitkrjfubh43jt19owanuezhntag5wh74c6ur5feuotpi73q8z
+// ----+------------+--------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------
+new
+  RevVault, rs(`rho:registry:insertSigned:ed25519`), uriOut,
+  rl(`rho:registry:lookup`), RevAddress(`rho:rev:address`), MakeMintCh, LockboxCh, EitherCh,
+  _makeVault,
+  _newVault,
+  _getOrCreate,
+  _revVault,
+  stdout(`rho:io:stdout`)
+in {
+  rl!(`rho:id:exunyijimapk7z43g3bbr69awqdz54kyroj9q43jgu3dh567fxsftx`, *MakeMintCh) |
+  rl!(`rho:id:9dsr55z1js13x346yhhecx66ns486i3yqf6jafrd9p9hdrrbxjqmyu`, *LockboxCh) |
+  rl!(`rho:id:j6trahbxycumerwpr5qype7j43b7eqh8auebwsa9nn68if47gswh73`, *EitherCh) |
+  for(@(_, MakeMint) <- MakeMintCh; @(_, Lockbox) <- LockboxCh; @(_, Either) <- EitherCh) {
+
+    new mintCh, vaultMapStore in {
+      vaultMapStore!({}) |
+      @MakeMint!(*mintCh) |
+      for (mint <- mintCh) {
+
+        //FIXME the result of createGenesisVault is not protected, nor is the call to it. Both should be.
+        contract RevVault(@"createGenesisVault", @ownerPubKeyBytes, @initialAmount, ret) = {
+          new createGenesisVault in {
+            _getOrCreate!(ownerPubKeyBytes, *createGenesisVault, *ret) |
+            for (retCh <- createGenesisVault) {
+
+              _makeVault!(ownerPubKeyBytes, initialAmount, *retCh)
+            }
+          }
+        } |
+
+        contract RevVault(@"findOrCreate", @ownerPubKeyBytes, ret) = {
+
+          new createLockedVault in {
+            _getOrCreate!(ownerPubKeyBytes, *createLockedVault, *ret) |
+            for (retCh <- createLockedVault) {
+
+              new vaultCh, lock in {
+                _makeVault!(ownerPubKeyBytes, 0, *vaultCh) |
+                @Either!("map <-", *vaultCh, *lock, *retCh) |
+                for (@vault, retCh <- lock) {
+                  @Lockbox!("lock", ownerPubKeyBytes, vault, *retCh)
+                }
+              }
+            }
+          }
+        } |
+
+        contract _getOrCreate(@key, constructor, ret) = {
+          for (@vaultMap <- vaultMapStore) {
```

Comment:
I think this is such a basic idiom and so easy to do with the just the language (without libs), that adding a lib contract would defeat the purpose. It might make sense if we find some substantial functionality to build on top of that, but as it is now - it's just a send + receive on a channel that only you know, and that stores at most 1 value. Doing anything around that requires the user to start using the 'method call convention', and a single 'method call' is as complex to set up as the whole idiom. So I really don't see any benefit for now (without a method call protocol built into rholang)